pkg/arn/arnlike.go

```diff
@@ -0,0 +1,131 @@
+package arn
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+	"unicode/utf8"
+)
+
+const (
+	arnDelimiter        = ":"
+	arnSectionsExpected = 6
+	arnPrefix           = "arn:"
+
+	// zero-indexed
+	sectionPartition = 1
+	sectionService   = 2
+	sectionRegion    = 3
+	sectionAccountID = 4
+	sectionResource  = 5
+
+	// errors
+	invalidPrefix   = "invalid prefix"
+	invalidSections = "not enough sections"
+)
+
+// ArnLike takes an ARN and returns true if it is matched by the pattern.
+// Each component of the ARN is matched individually as per
+// https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_condition_operators.html#Conditions_ARN
+func ArnLike(arn, pattern string) (bool, error) {
+	// "parse" the input arn into sections
+	arnSections, err := parse(arn)
+	if err != nil {
+		return false, fmt.Errorf("Could not parse input arn: %v", err)
+	}
+	patternSections, err := parse(pattern)
+	if err != nil {
+		return false, fmt.Errorf("Could not parse ArnLike string: %v", err)
+	}
+
+	// Tidy regexp special characters. Escape the ones not used in ArnLike.
+	// Replace multiple * with .* - we're assuming `\` is not allowed in ARNs
+	preparePatternSections(patternSections)
+
+	for index := range arnSections {
+		patternGlob, err := regexp.Compile(patternSections[index])
+		if err != nil {
+			return false, fmt.Errorf("Could not parse %s: %v", patternSections[index], err)
+		}
+
+		if !patternGlob.MatchString(arnSections[index]) {
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+// parse is a copy of arn.Parse from the AWS SDK but represents the ARN as []string
+func parse(input string) ([]string, error) {
+	if !strings.HasPrefix(input, arnPrefix) {
+		return nil, fmt.Errorf(invalidPrefix)
+	}
+	arnSections := strings.SplitN(input, arnDelimiter, arnSectionsExpected)
+	if len(arnSections) != arnSectionsExpected {
+		return nil, fmt.Errorf(invalidSections)
+	}
+
+	return arnSections, nil
+}
+
+// preparePatternSections goes through each section of the arnLike slice and escapes any meta characters, except for
+// `*` and `?` which are replaced by `.*` and `.?` respectively. ^ and $ are added as we require an exact match
+func preparePatternSections(arnLikeSlice []string) {
+	for index, section := range arnLikeSlice {
+		quotedString := quoteMeta(section)
+		arnLikeSlice[index] = `^` + quotedString + `$`
+	}
+}
+
+// the below is copied from regexp.QuoteMeta to escape everything but `?` and change `*` to `.*`
+// regexp.QuoteMeta
+
+// Bitmap used by func special to check whether a character needs to be escaped.
+var specialBytes [16]byte
+
+// special reports whether byte b needs to be escaped by QuoteMeta.
+func special(b byte) bool {
+	return b < utf8.RuneSelf && specialBytes[b%16]&(1<<(b/16)) != 0
+}
+
+func init() {
+	for _, b := range []byte(`\.+()|[]{}^$`) {
+		specialBytes[b%16] |= 1 << (b / 16)
+	}
+}
```

Comment:
Would an easier-to-read (and easier to vet for security purposes) approach be to do something like this instead of the bitmap approach?

```go
var specialChars []byte(`\.+()|[]{}^$`)

func containsSpecial(subject string) bool {
	for _, sc := range specialChars {
		if strings.ContainsRune(subject, sc) {
			return true
		}
	}
	return false
}
```

I mean, this isn't in the performance-sensitive code path, right? This is just used when preparing the matching regexes...

Reply:
That implementation was copied from regexp.QuoteMeta. While it's not performance critical, I do feel like it's going to be called a lot of times.

Anyway I've simplified the implementation considerably